app/Commands/Dev/Tree/Compile/Base.hs

```diff
@@ -0,0 +1,93 @@
+module Commands.Dev.Tree.Compile.Base where
+
+import Commands.Base
+import Commands.Dev.Tree.Compile.Options
+import Commands.Extra.Compile qualified as Compile
+import Juvix.Compiler.Asm.Pretty qualified as Asm
+import Juvix.Compiler.Backend qualified as Backend
+import Juvix.Compiler.Backend.C qualified as C
+import Juvix.Compiler.Nockma.Pretty qualified as Nockma
+import Juvix.Compiler.Tree.Data.InfoTable qualified as Tree
+
+data PipelineArg = PipelineArg
+  { _pipelineArgOptions :: CompileOptions,
+    _pipelineArgFile :: Path Abs File,
+    _pipelineArgTable :: Tree.InfoTable
+  }
+
+getEntry :: (Members '[Embed IO, App, TaggedLock] r) => PipelineArg -> Sem r EntryPoint
+getEntry PipelineArg {..} = do
+  ep <- getEntryPoint (AppPath (preFileFromAbs _pipelineArgFile) True)
+  return $
+    ep
+      { _entryPointTarget = getTarget (_pipelineArgOptions ^. compileTarget),
+        _entryPointDebug = _pipelineArgOptions ^. compileDebug,
+        _entryPointUnsafe = _pipelineArgOptions ^. compileUnsafe,
+        _entryPointOptimizationLevel = fromMaybe defaultOptLevel (_pipelineArgOptions ^. compileOptimizationLevel),
+        _entryPointInliningDepth = _pipelineArgOptions ^. compileInliningDepth
+      }
+  where
+    getTarget :: CompileTarget -> Backend.Target
+    getTarget = \case
+      TargetWasm32Wasi -> Backend.TargetCWasm32Wasi
+      TargetNative64 -> Backend.TargetCNative64
+      TargetGeb -> Backend.TargetGeb
+      TargetVampIR -> Backend.TargetVampIR
+      TargetCore -> Backend.TargetCore
+      TargetAsm -> Backend.TargetAsm
+      TargetTree -> Backend.TargetTree
+      TargetNockma -> Backend.TargetNockma
+
+    defaultOptLevel :: Int
+    defaultOptLevel
+      | _pipelineArgOptions ^. compileDebug = 0
+      | otherwise = defaultOptimizationLevel
+
+runCPipeline ::
+  forall r.
+  (Members '[Embed IO, App, TaggedLock] r) =>
+  PipelineArg ->
+  Sem r ()
+runCPipeline pa@PipelineArg {..} = do
+  entryPoint <- getEntry pa
+  C.MiniCResult {..} <- getRight (run (runReader entryPoint (runError (treeToMiniC _pipelineArgTable :: Sem '[Error JuvixError, Reader EntryPoint] C.MiniCResult))))
```

Comment:
this line would look nicer with the `f . g $ x` shape.
Also, we usually put type annotations next to the run function. Something like
```suggestion
  C.MiniCResult {..} <- getRight 
        . run 
        . runReader entryPoint 
        . runError @JuvixError
        $ treeToMiniC _pipelineArgTable
```

Reply:
done

Reply:
This was copied and modified from `Commands/Dev/Core/Compile/Base.hs`, which perhaps should also be changed in a future refactor.